SimPEG/directives/pgi_directives.py

```diff
@@ -27,108 +26,142 @@
+        # current iteration doesn't match the update rate
+        if self.opt.iter == 0 or self.opt.iter % self.update_rate != 0:
+            return None
 
+        # Get current model
+        m = self.invProb.model
+        modellist = self.pgi_regularization.wiresmap * m
+        model = np.c_[
+            [a * b for a, b in zip(self.pgi_regularization.maplist, modellist)]
+        ].T
+
+        # Update the GMM parameters if required
+        gmmref = self.pgi_regularization.gmmref
+        if self.update_gmm:
+            # Determine the class of the new gmm
+            if isinstance(gmmref, GaussianMixtureWithNonlinearRelationships):
+                new_gmm_class = GaussianMixtureWithNonlinearRelationshipsWithPrior
+            elif isinstance(gmmref, WeightedGaussianMixture):
+                new_gmm_class = GaussianMixtureWithPrior
             else:
-                clfupdate = copy.deepcopy(self.pgi_reg.gmmref)
-
-            self.pgi_reg.gmm = clfupdate
-            membership = self.pgi_reg.gmm.predict(model)
-
-            if self.fixed_membership is not None:
-                membership[self.fixed_membership[:, 0]] = self.fixed_membership[:, 1]
+                raise ValueError(
+                    f"Invalid class '{gmmref.__class__}' for the Gaussian Mixture"
+                    " in the PGI regularization."
+                )
+            new_gmm = new_gmm_class(
+                gmmref=gmmref,
+                zeta=self.zeta,
+                kappa=self.kappa,
+                nu=self.nu,
+                verbose=self.verbose,
+                prior_type="semi",
+                update_covariances=self.update_covariances,
+                max_iter=self.pgi_regularization.gmm.max_iter,
+                n_init=self.pgi_regularization.gmm.n_init,
+                reg_covar=self.pgi_regularization.gmm.reg_covar,
+                weights_init=self.pgi_regularization.gmm.weights_,
+                means_init=self.pgi_regularization.gmm.means_,
+                precisions_init=self.pgi_regularization.gmm.precisions_,
+                random_state=self.pgi_regularization.gmm.random_state,
+                tol=self.pgi_regularization.gmm.tol,
+                verbose_interval=self.pgi_regularization.gmm.verbose_interval,
+                warm_start=self.pgi_regularization.gmm.warm_start,
+                fixed_membership=self.fixed_membership,
+            )
+            new_gmm = new_gmm.fit(model)
+        else:
+            new_gmm = copy.deepcopy(gmmref)
+
+        # Update the GMM in the PGI regularization
+        self.pgi_regularization.gmm = new_gmm
+        # Compute membership of the current model
+        membership = self.pgi_regularization.gmm.predict(model)
+        # Override the membership if some cells have fixed membership
+        if self.fixed_membership is not None:
+            membership[self.fixed_membership[:, 0]] = self.fixed_membership[:, 1]
```

Comment:
After updating the GMM, I saw that the membership number could change. I mean that a rock unit associated with a membership `0` in the reference GMM, could be related to the membership `1` in the updated GMM.
Therefore, the membership indices in the `fixed_membership` would be wrong and produce an unwanted result.

I think we should test and fix this.

Reply:
> After updating the GMM, I saw that the membership number could change. I mean that a rock unit associated with a membership 0 in the reference GMM, could have a the membership 1 in the updated GMM.

This is supposed to be dealt with by the `order_cluster` function. But it might have some mild assumptions like "the highest proportions is for the background". But in my tests it was very robust. See for example my thesis, Figure 6.5, where a range of proportions are tried, with even some cases where the background weight is very very small.